pkg/storage/backup_compaction_iterator_test.go

```diff
@@ -0,0 +1,192 @@
+		{input: "b1", expectedNextKey: "b1", asOf: "1"},
+		{input: "b2b1", expectedNextKey: "b1", asOf: "1"},
+
+		// Double skip within keys.
+		{input: "b3b2b1", expectedNextKey: "b1", asOf: "1"},
+
+		// Double skip across keys.
+		{input: "b2c2c1", expectedNextKey: "c1", asOf: "1"},
+
+		// Ensure next key captures at most one mvcc key per key after an asOf skip.
+		{input: "b3c2c1", expectedNextKey: "c2", asOf: "2"},
+
+		// Ensure an AOST 'next' takes precedence over a tombstone 'nextkey'.
+		{input: "b2Xb1c1", expectedNextKey: "b2Xc1", asOf: ""},
+		{input: "b2Xb1c1", expectedNextKey: "b1c1", asOf: "1"},
+
+		// Ensure clean iteration over double tombstone.
+		{input: "a1Xb2Xb1c1", expectedNextKey: "a1Xb2Xc1", asOf: ""},
+		{input: "a1Xb2Xb1c1", expectedNextKey: "a1Xb1c1", asOf: "1"},
+
+		// Ensure tombstone is skipped after an AOST skip.
+		{input: "b3c2Xc1d1", expectedNextKey: "c1d1", asOf: "1"},
+
+		// Ensure key before delete tombstone gets read if under AOST.
+		{input: "b2b1Xc1", expectedNextKey: "b2c1", asOf: ""},
+		{input: "b2b1Xc1", expectedNextKey: "b2c1", asOf: "2"},
+	}
+
+	for i, test := range tests {
+		name := fmt.Sprintf("Test %d: %s, AOST %s", i, test.input, test.asOf)
+		t.Run(name, func(t *testing.T) {
+			batch := pebble.NewBatch()
+			defer batch.Close()
+			populateBatch(t, batch, test.input)
+			iter, err := batch.NewMVCCIterator(context.Background(), MVCCKeyAndIntentsIterKind, IterOptions{UpperBound: roachpb.KeyMax})
+			require.NoError(t, err)
+			defer iter.Close()
+
+			subtests := []iterSubtest{
+				{"NextKey", test.expectedNextKey, SimpleMVCCIterator.NextKey},
+			}
+			for _, subtest := range subtests {
+				t.Run(subtest.name, func(t *testing.T) {
+					asOf := hlc.Timestamp{}
+					if test.asOf != "" {
+						asOf.WallTime = int64(test.asOf[0])
+					}
+					it := NewBackupCompactionIterator(iter, asOf)
+					iterateSimpleMVCCIterator(t, it, subtest)
+				})
+			}
+		})
+	}
+}
+
+func TestBackupCompactionIteratorSeek(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	pebble, err := Open(context.Background(), InMemory(),
+		cluster.MakeTestingClusterSettings(), CacheSize(1<<20 /* 1 MiB */))
+	require.NoError(t, err)
+	defer pebble.Close()
+
+	tests := []struct {
+		input    string
+		seekKey  string
+		expected string
+		asOf     string
+	}{
+		// Ensure vanilla seek works.
+		{"a1b1", "a1", "a1", ""},
+
+		// Ensure seek always returns the latest key of an MVCC key.
+		{"a2a1b1", "a1", "b1", ""},
+		{"a2a1b1", "a1", "b1", "2"},
+		{"a2a1b1", "a1", "a1", "1"},
+
+		// Ensure out of bounds seek fails gracefully.
+		{"a1", "b1", "notOK", ""},
+
+		// Ensure the asOf timestamp moves the iterator during a seek.
+		{"a2a1", "a2", "a1", "1"},
+		{"a2b1", "a2", "b1", "1"},
+
+		// Ensure seek will return on a tombstone if it is the latest key.
+		{"a3Xa1b1", "a3", "a3X", ""},
+		{"a3Xa1c2Xc1", "b1", "c2X", ""},
+
+		// Ensure seek does not return on a key shadowed by a tombstone.
```

Comment:
this seek behavior is unrelated to the tombstone, we always seek to b1 because a2 and a1 are not the latest key given the aost. i'd rephrase the comment here.